test/calibration/test_setup_library.py

```diff
@@ -119,3 +123,100 @@ def test_setup_partial_gates(self):
             FixedFrequencyTransmon(basis_gates=["x", "bswap"])
+
+    def test_serialization(self):
+        """Test the serialization of the object."""
+
+        lib1 = FixedFrequencyTransmon(
+            basis_gates=["x", "sy"],
+            default_values={"duration": 320},
+            use_drag=False,
+            link_parameters=False,
+        )
+
+        lib2 = FixedFrequencyTransmon.from_config(lib1.config)
+
+        self.assertEqual(lib2.basis_gates, lib1.basis_gates)
+
+        # Note: we convert to string since the parameters prevent a direct comparison.
+        self.assertTrue(self._test_library_equivalence(lib1, lib2))
+
+        # Test that the extra args are properly accounted for.
+        lib3 = FixedFrequencyTransmon(
+            basis_gates=["x", "sy"],
+            default_values={"duration": 320},
+            use_drag=True,
+            link_parameters=False,
+        )
+
+        self.assertFalse(self._test_library_equivalence(lib1, lib3))
+
+    def test_json_serialization(self):
+        """Test that the library can be serialized using JSon."""
+
+        lib1 = FixedFrequencyTransmon(
+            basis_gates=["x", "sy"],
+            default_values={"duration": 320},
+            use_drag=False,
+            link_parameters=False,
+        )
+
+        # Test that serialization fails without the right encoder
+        with self.assertRaises(TypeError):
+            json.dumps(lib1)
+
+        # Test that serialization works with the proper library
+        lib_data = json.dumps(lib1, cls=ExperimentEncoder)
+        lib2 = json.loads(lib_data, cls=ExperimentDecoder)
+
+        self.assertTrue(self._test_library_equivalence(lib1, lib2))
+
+    def test_hash_warn(self):
+        """Test that a warning is raised when the hash of the library is different.
+
+        This test mimics the behaviour of the following workflow:
+        1. A user serializes a library.
+        2. Changes to the class of the library are made.
+        3. The user deserializes the library with the changed class.
+        4. A warning is raised since the class definition has changed.
+        """
+
+        lib1 = FixedFrequencyTransmon()
+        lib_data = json.dumps(lib1, cls=ExperimentEncoder)
+        lib2 = json.loads(lib_data, cls=ExperimentDecoder)
+
+        self.assertTrue(self._test_library_equivalence(lib1, lib2))
+
+        # stash method build schedules to avoid other tests from failing
+        build_schedules = FixedFrequencyTransmon._build_schedules
+
+        def _my_build_schedules():
+            """A dummy function to change the class behaviour."""
+            pass
+
+        # Change the schedule behaviour
+        FixedFrequencyTransmon._build_schedules = _my_build_schedules
+
+        with self.assertWarns(UserWarning):
+            json.loads(lib_data, cls=ExperimentDecoder)
+
+        # Restore the method for future tests.
+        FixedFrequencyTransmon._build_schedules = build_schedules
```

Comment:
Could you subclass `FixedFrequencyTransmon` with a modified `_build_schedules` instead? I am worried that if there is a bug in the json decoding weird things will happen in other tests if this line doesn't get run.

Reply:
Good suggestion. See here: 3f803bf and here 46bba21